File: IDATT2106 Systemutvikling 2 med smidig prosjekt/backend/src/test/java/com/ntnu/gidd/service/UserServiceImplTest.java
package com.ntnu.gidd.service;

import com.ntnu.gidd.dto.User.UserDto;
import com.ntnu.gidd.dto.User.UserRegistrationDto;
import com.ntnu.gidd.factories.UserFactory;
import com.ntnu.gidd.model.User;
import com.ntnu.gidd.repository.PostRepository;
import com.ntnu.gidd.repository.PasswordResetTokenRepository;
import com.ntnu.gidd.repository.TrainingLevelRepository;
import com.ntnu.gidd.repository.UserRepository;
import com.ntnu.gidd.service.Comment.CommentService;
import com.ntnu.gidd.service.Email.EmailService;
import com.ntnu.gidd.service.Registration.RegistrationService;
import com.ntnu.gidd.service.User.UserServiceImpl;
import com.ntnu.gidd.util.ContextAwareModelMapper;
import com.ntnu.gidd.utils.JpaUtils;
import com.querydsl.core.types.Predicate;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.modelmapper.ModelMapper;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
public class UserServiceImplTest {

    @InjectMocks
    private UserServiceImpl userService;

    @Mock
    private UserRepository userRepository;

    @Mock
    private RegistrationService registrationService;

    @Mock
    private CommentService commentService;

    @Mock
    PostRepository postRepository;

    private EmailService emailService;

    @Mock
    private PasswordResetTokenRepository passwordResetTokenRepository;

    @Mock
    private TrainingLevelRepository trainingLevelRepository;

    @Mock
    BCryptPasswordEncoder encoder;

    ModelMapper modelMapper = new ContextAwareModelMapper();

    private User user;
    private User secondUser;
    private List<User> usersExpected;
    private UserRegistrationDto userRegistrationDto;
    private Predicate predicate;
    private Pageable pageable;

    @BeforeEach
    void setUp() throws Exception {
        userService = new UserServiceImpl(encoder,
                modelMapper,
                userRepository,
                emailService,
                passwordResetTokenRepository, registrationService,
                commentService,
                postRepository,
                trainingLevelRepository);
        user = new UserFactory().getObject();
        userRegistrationDto = modelMapper.map(user, UserRegistrationDto.class);
        userRepository.save(user);

        predicate = JpaUtils.getEmptyPredicate();
        pageable = JpaUtils.getDefaultPageable();
    }

    @AfterEach
    void cleanUp() {
        userRepository.deleteAll();
    }

    @Test
    void testGetUserDtoByEmailReturnsUserDto() {
        when(userRepository.findByEmail(user.getEmail())).thenReturn(Optional.of(user));
        UserDto userDtoFound = userService.getUserDtoByEmail(user.getEmail());
        assertThat(userDtoFound.getEmail()).isEqualTo(user.getEmail());
    }

    @Test
    void testDeleteUserDeletesUser() throws Exception{
        secondUser = new UserFactory().getObject();
        userRepository.save(secondUser);
        usersExpected = List.of(user, secondUser);
        when(userRepository.findAll(any(Predicate.class), any(PageRequest.class))).thenReturn(new PageImpl<>(usersExpected, pageable, usersExpected.size()));
        Page<UserDto> usersFound = userService.getAll(predicate, pageable);

        assertThat(usersFound.stream().findFirst().get().getEmail()).
                isEqualTo(usersExpected.get(0).getEmail());
        assertThat(usersFound.getTotalElements())
                .isEqualTo(Long.valueOf(usersExpected.size()));

        when(userRepository.findByEmail(user.getEmail())).thenReturn(Optional.of(user));
        doNothing().when(userRepository).deleteById(user.getId());
        doNothing().when(registrationService).deleteAllRegistrationsWithUsername(user.getEmail());
        doNothing().when(commentService).deleteAllCommentsOnUser(user.getEmail());
        doNothing().when(postRepository).deletePostsByCreator(user);
        userService.deleteUser(user.getEmail());

        List<User> usersExpectedAfterDelete = List.of(secondUser);
        when(userRepository.findAll(any(Predicate.class), any(PageRequest.class))).thenReturn(new PageImpl<>(usersExpectedAfterDelete, pageable, usersExpected.size() - 1));
        Page<User> usersFound2 = userRepository.findAll(predicate, pageable);

        assertThat(usersFound2.getTotalElements())
                .isEqualTo(Long.valueOf(usersExpected.size() - 1));
        assertThat(usersFound2.stream().findFirst().get().getEmail()).
                isEqualTo(usersExpected.get(1).getEmail());
    }
}

File: IDATT2105 Full-stack applikasjonsutvikling/Rombestilling/backend/src/test/kotlin/ntnu/idatt2105/user/service/UserServiceImplTest.kt
package ntnu.idatt2105.user.service

import io.github.serpro69.kfaker.Faker
import ntnu.idatt2105.core.config.ModelMapperConfig
import ntnu.idatt2105.core.exception.ApplicationException
import ntnu.idatt2105.core.mailer.MailService
import ntnu.idatt2105.factories.UserFactory
import ntnu.idatt2105.security.repository.PasswordResetTokenRepository
import ntnu.idatt2105.user.dto.DetailedUserDto
import ntnu.idatt2105.user.dto.UserDto
import ntnu.idatt2105.user.model.User
import ntnu.idatt2105.user.repository.RoleRepository
import ntnu.idatt2105.user.repository.UserRepository
import ntnu.idatt2105.util.JpaUtils
import org.assertj.core.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.mockito.ArgumentMatchers.any
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension
import org.modelmapper.ModelMapper
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import java.util.*
import java.util.stream.Stream

@ExtendWith(MockitoExtension::class)
internal class UserServiceImplTest {

    private lateinit var userService: UserService

    @Mock
    private lateinit var userRepository: UserRepository

    @Mock
    private lateinit var passwordEncoder: BCryptPasswordEncoder

    @Mock
    private lateinit var mailService: MailService

    @Mock
    private lateinit var passwordResetTokenRepository: PasswordResetTokenRepository

    @Mock
    private lateinit var roleRepository: RoleRepository

    private lateinit var modelMapper: ModelMapper

    private lateinit var userFactory: UserFactory

    private lateinit var defaultPageable: Pageable

    @BeforeEach
    fun setUp() {
        modelMapper = ModelMapperConfig().modelMapper()
        userFactory = UserFactory()
        userService = UserServiceImpl(
            userRepository = userRepository,
            modelMapper = modelMapper,
            passwordEncoder = passwordEncoder,
            mailService = mailService,
            passwordResetTokenRepository = passwordResetTokenRepository,
            roleRepository = roleRepository
        )

        defaultPageable = JpaUtils().getDefaultPageable()
    }

    @Test
    fun `test get users returns all users`() {
        val users = IntRange(1, 4).map { userFactory.`object` }
        val expectedPage: Page<User> = PageImpl(users, JpaUtils().getDefaultPageable(), users.size.toLong())
        val predicate = JpaUtils().getEmptyPredicate()
        `when`(userRepository.findAll(predicate, defaultPageable))
            .thenReturn(expectedPage)

        val actualUsers = userService.getUsers(defaultPageable, predicate)

        assertThat(actualUsers.size).isEqualTo(expectedPage.size)
        assertThat(actualUsers.content).doesNotContainNull()
    }

    @Test
    fun `test get user when user not found throws exception`() {
        `when`(userRepository.findById(any(UUID::class.java)))
            .thenThrow(ApplicationException.EntityNotFoundException("Ignored"))

        assertThatExceptionOfType(ApplicationException.EntityNotFoundException::class.java)
            .isThrownBy { userService.getUser(UUID.randomUUID(), UserDto::class.java) }
    }

    @Test
    fun `test get user when user exists returns the user`() {
        val user = userFactory.`object`
        `when`(userRepository.findById(user.id))
            .thenReturn(Optional.of(user))

        val actualUser = userService.getUser(user.id, UserDto::class.java)

        assertThat(actualUser.id).isEqualTo(user.id)
    }

    @ParameterizedTest
    @MethodSource("userDtoClassProvider")
    fun `test get user maps to correct dto class`(dtoClass: Class<*>) {
        val user = userFactory.`object`
        `when`(userRepository.findById(user.id))
            .thenReturn(Optional.of(user))

        val actualUser = userService.getUser(user.id, mapTo = dtoClass)

        assertThat(actualUser).isExactlyInstanceOf(dtoClass)
    }

    companion object {
        @JvmStatic
        private fun userDtoClassProvider() =
            Stream.of(
                Arguments.of(UserDto::class.java),
                Arguments.of(DetailedUserDto::class.java),
            )
    }

    @Test
    fun `test update user updates user with the given id`() {
        val user = userFactory.`object`
        `when`(userRepository.findById(user.id))
            .thenReturn(Optional.of(user))

        user.email = Faker().internet.email()

        `when`(userRepository.save(any(User::class.java)))
            .thenReturn(user)

        val updateUserRequest = modelMapper.map(user, UserDto::class.java)
        val actualUser = userService.updateUser(user.id, updateUserRequest)

        assertThat(actualUser.email).isEqualTo(user.email)
        verify(userRepository, times(1)).save(user)
    }

    @Test
    fun `test update user when not found throws exception`() {
        `when`(userRepository.findById(any(UUID::class.java)))
            .thenThrow(ApplicationException.EntityNotFoundException("Ignored"))

        val updateUserRequest = modelMapper.map(userFactory.`object`, UserDto::class.java)

        assertThatExceptionOfType(ApplicationException.EntityNotFoundException::class.java)
            .isThrownBy { userService.updateUser(UUID.randomUUID(), updateUserRequest) }
    }

    @Test
    fun `test delete user when user not found throws exception`() {
        `when`(userRepository.findById(any(UUID::class.java)))
            .thenThrow(ApplicationException.EntityNotFoundException("Ignored"))

        assertThatExceptionOfType(ApplicationException.EntityNotFoundException::class.java)
            .isThrownBy { userService.deleteUser(UUID.randomUUID()) }
    }

    @Test
    fun `test delete user when user exists deletes user with given id`() {
        val user = userFactory.`object`
        `when`(userRepository.findById(user.id))
            .thenReturn(Optional.of(user))

        userService.deleteUser(user.id)

        verify(userRepository, times(1)).delete(user)
    }
}
